### ISSUE TITLE
[HOLD for payment 2023-04-10] [$1000] [Regression] Request Money - Amount field not auto focused if suddenly loose focus

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#16511]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/16394

## Action Performed:
1. Launch expensify app
2. Login with any account
3. Tap on the FAB button -> Send money
4. Enter numbers
5. Click outside of numpad
6. Enter numbers

## Expected Result:
Autofocus Amount field if suddenly loose focus

## Actual Result:
Amount field not auto focused if suddenly loose focus

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.89.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/227730584-334957a6-c2e6-4b3d-8e2e-2baa715af479.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01dbf046ada3294608</li>
        <li>Upwork Job ID: 1640296677002747904</li>
        <li>Last Price Increase: 2023-03-27</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sonialiap (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by luacmartins** [comment]
@mountiny @0xmiroslav @narefyev91 since you worked on https://github.com/Expensify/App/pull/16394. Tapping outside the input doesn't allow for further inputs

**Comment by mountiny** [comment]
Thanks for the bump here, I will take this one on with @narefyev91 again. Changes to this page open a can of worms sorry

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01dbf046ada3294608

**Comment by MelvinBot** [comment]
Current assignee @sonialiap is eligible for the External assigner, not assigning anyone new.

**Comment by narefyev91** [comment]
I'm Nikolay from Callstack - and i will take a look on this one

**Comment by MelvinBot** [comment]
📣 @narefyev91 You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by mountiny** [comment]
@0xmiroslav sorry it wasnt clear from you comment on the PR, your suggestion seemed to introduce a [different](https://github.com/Expensify/App/pull/16394#issuecomment-1480735810) regression.

Let's try to properly investigate this page and get the best cross platform solution, I truest in @0xmiroslav and @narefyev91 skills here to get to the best solution together 🤝 thanks 🙇

**Comment by narefyev91** [comment]
@mountiny @0xmiroslav As a workaround seems we need to have custom event handler to detect onMouseDown event and only preventDefault() if we click on main wrapper, but not on input (to allow user to move cursor)
```
    /**
     * Listens for mouse down event and applies the action
     *
     * @param {Object} e
     */
    handleMouseDown(e) {
        if (e.target.nodeName !== 'DIV') { return; }
        e.preventDefault();
    }
```

https://user-images.githubusercontent.com/28352309/227934461-f73056bc-a039-4e24-bf23-9b716346565d.mov

**Comment by mountiny** [comment]
Currency selector will still work fine?

**Comment by narefyev91** [comment]
@mountiny yup

https://user-images.githubusercontent.com/28352309/227936140-6f158aa3-69ad-437a-8772-266243f091c7.mov

**Comment by mountiny** [comment]
Nice, curious for @0xmiroslav thoughts!

Thanks!

**Comment by mountiny** [comment]
> I will stick with first approach - flickering issue should be fixed globally

Is the flickering issue with TextInput resolved by enabling Fabric

**Comment by narefyev91** [comment]
@mountiny hard to say - I can't find issue in github to check why those changes were implemented - but seems after that changes were merged - issue started to be tracked - and in case it's direct css manipulation - not sure if something will be changed after fabric will be enabled, but i can be wrong
<img width="1727" alt="Screenshot 2023-03-27 at 19 06 12" src="https://user-images.githubusercontent.com/28352309/227999795-dccf04bb-0415-479d-a79f-eb81d2292f2f.png">

**Comment by mountiny** [comment]
Quite a tough nut to crack this one. What if we take a step back further looking at how this page is implemented, is there something which gives us these limitations but we could maybe refactor for something better. We will go to hooks soon so we will refactor this nonetheless

**Comment by narefyev91** [comment]
Sure i will take a look on different angle

**Comment by ahmedGaber93** [comment]
replace `View` warpper with `Pressable` and focus input again when `onPressIn` will focus input again after losing focus


```
import { Pressable } from 'react-native';

<Pressable
      onPressIn={() => this.textInput.focus()}
```

**Comment by MelvinBot** [comment]
📣 @ahmedGaber93! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by narefyev91** [comment]
@mountiny @0xmiroslav The idea from ahmedGaber93 is also partially works - to switch to onResponderStart={e => e.preventDefault()} this one - but again it still the same workaround which we discussed yesterday. Based on my view - after checking deeper the logic around the screens: if we simplify what we currently have - Container -> Input -> Button (NumPad). In any case for web we need to control click outside based on event (onMouseDown,  onResponderStart, onBlur) - all this workflows are correct - and the only issue we facing now in most cases - it's flickering issue. To avoid this issue we need to remove clicking event outside of input. And we coming again to the same proposals we have yesterday. I think we need to choose one of the approach and move forward

**Comment by mountiny** [comment]
Thank you!

@0xmiroslav @narefyev91 which one would you prefer

**Comment by mountiny** [comment]
I think both will bring in the flickering issue, but lets just go with what is more simple which (to me) feels like the click outside mouseDown approach from @narefyev91

Does that work @0xmiroslav or any strong argument why not to go that way

**Comment by mountiny** [comment]
Thanks, those are good arguments for not going that way.

@narefyev91 any arguments for not going the onBlur way @0xmiroslav proposed?

**Comment by narefyev91** [comment]
@mountiny Ok sure - we can try onBlur
@0xmiroslav can you please explain what should we do - as i understand you - we need to add logic for TextInput onBlur - when onBlur fires we need to check what exactly triggered to loose textinput focus, and if that was done by clicking on dark area we need to focus again? am I right?

**Comment by narefyev91** [comment]
@0xmiroslav ok - after some work around i extract this to onBlur
```
                        onBlur={(event) => {
                            const relatedTargetId = lodashGet(event, 'nativeEvent.relatedTarget.id');
                            if (!relatedTargetId) {
                                this.focusTextInput();
                            }
                        }}
```
Also i add 2 nativeIDs for Button in the bottom, and for currency select as well
Based on this - if user clicks anywhere on the dark - event will not have `relatedTarget.id` - which will be indication that user clicks somewhere but not on currency selector/textinput itself/button in the bottom and we will move back focus to input
Let me know if that works for us

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.93-4 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16673


If no regressions arise, payment will be issued on 2023-04-10. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@0xmiroslav / @mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@0xmiroslav / @mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@0xmiroslav / @mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sonialiap] Determine if we should create a regression test for this bug.
- [ ] [@0xmiroslav] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sonialiap] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mountiny** [comment]
@0xmiroslav would you be able to help with the Checklist please? Then we can pay out

**Comment by mountiny** [comment]
Thank you! I think Sonia is ooo so I will ask someone to help with this.

**Comment by Christinadobrzyn** [comment]
hey @narefyev91 can you please provide your Upwork profile so I can hire you for this job?

**Comment by Christinadobrzyn** [comment]
Ah thank you for confirming @0xmiroslav!

**Comment by hayata-suenaga** [comment]
I'm coming to this issue from [another issue issue](https://github.com/Expensify/App/issues/36862#issuecomment-1969051706).

The `event.preventDefault();` was added in [the the PR](https://github.com/Expensify/App/pull/16673/files) to fix this issue. However, the original issue doesn't seem to be a problem. Isn't it an expected behavior that the text input looses focus if you click outside of it?

**Comment by hayata-suenaga** [comment]
decided to go with [this fix](https://github.com/Expensify/App/issues/36862#issuecomment-2020268956)
